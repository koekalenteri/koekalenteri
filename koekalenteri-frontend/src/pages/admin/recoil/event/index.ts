export * from "./currentEvent"
export * from "./eventClassAtom"
export * from "./eventSelector"
