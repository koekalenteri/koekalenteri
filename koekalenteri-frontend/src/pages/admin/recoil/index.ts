export * from './emailTemplates'
export * from './events'
export * from './officials'
export * from './organizers'
export * from './registrations'
