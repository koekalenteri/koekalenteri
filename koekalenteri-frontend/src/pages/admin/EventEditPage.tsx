import { useCallback, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { useNavigate, useParams } from 'react-router-dom'
import { Event } from 'koekalenteri-shared/model'
import { useSnackbar } from 'notistack'
import { useRecoilState, useRecoilValue, useResetRecoilState } from 'recoil'

import { Path } from '../../routeConfig'
import { hasChanges } from '../../utils'
import { activeEventTypesSelector, activeJudgesSelector, eventTypeClassesAtom } from '../recoil'

import EventForm from './eventEditPage/EventForm'
import {
  adminEventSelector,
  adminUserOrganizersSelector,
  editableEventByIdAtom,
  officialsAtom,
  useAdminEventActions,
} from './recoil'

export default function EventEditPage() {
  const { id: eventId = '' } = useParams()
  const { t } = useTranslation()
  const { enqueueSnackbar } = useSnackbar()
  const navigate = useNavigate()
  const activeEventTypes = useRecoilValue(activeEventTypesSelector)
  const activeJudges = useRecoilValue(activeJudgesSelector)
  const eventTypeClasses = useRecoilValue(eventTypeClassesAtom)
  const officials = useRecoilValue(officialsAtom)
  const organizers = useRecoilValue(adminUserOrganizersSelector)

  const actions = useAdminEventActions()
  const storedEvent = useRecoilValue(adminEventSelector(eventId))
  const [event, setEvent] = useRecoilState(editableEventByIdAtom(eventId))
  const resetEvent = useResetRecoilState(editableEventByIdAtom(eventId))
  const [changes, setChanges] = useState<boolean>(hasChanges(storedEvent, event))

  const handleChange = useCallback(
    (newState: Event) => {
      setChanges(hasChanges(storedEvent, newState))
      setEvent(newState)
    },
    [setEvent, storedEvent]
  )

  const handleSave = useCallback(async () => {
    if (!event) {
      return
    }
    await actions.save(event)
    resetEvent()
    navigate(Path.admin.events)
    enqueueSnackbar(t(`event.states.${event?.state || 'draft'}`, { context: 'save' }), { variant: 'info' })
  }, [actions, enqueueSnackbar, event, navigate, resetEvent, t])

  const handleCancel = useCallback(async () => {
    resetEvent()
    navigate(Path.admin.events)
  }, [navigate, resetEvent])

  if (!event) {
    return <div>event not found</div>
  }

  return (
    <EventForm
      event={event}
      changes={changes}
      eventTypes={activeEventTypes.map((et) => et.eventType)}
      eventTypeClasses={eventTypeClasses}
      judges={activeJudges}
      officials={officials}
      organizers={organizers}
      onChange={handleChange}
      onSave={handleSave}
      onCancel={handleCancel}
    />
  )
}
