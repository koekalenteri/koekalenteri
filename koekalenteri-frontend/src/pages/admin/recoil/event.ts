import { EventEx } from "koekalenteri-shared/model"
import { atom, DefaultValue, selector, selectorFamily } from "recoil"

import { getEvent, getEvents } from "../../../api/event"
import { unique, uniqueDate } from "../../../utils"
import { logEffect, storageEffect } from "../../recoil/effects"

export interface DecoratedEvent extends EventEx {
  uniqueClasses: string[]
  uniqueClassDates: (eventClass: string) => Date[]
}

export const adminEventIdAtom = atom<string | undefined>({
  key: 'adminEventId',
  default: undefined,
  effects: [
    logEffect,
    storageEffect,
  ],
})

export const currentAdminEvent = selector({
  key: 'currentAdminEvent',
  get: ({ get }) => {
    const eventId = get(adminEventIdAtom)
    return eventId ? get(adminEventSelector(eventId)) : undefined
  },
  set: ({ set }, newValue) => {
    if (!newValue || newValue instanceof DefaultValue) {
      return
    }
    set(adminEventSelector(newValue.id), newValue)
  },
})

function decorateEvent(event: EventEx): DecoratedEvent {
  return {
    ...event,
    uniqueClasses: unique(event.classes.map(c => c.class)),
    uniqueClassDates: (eventClass: string) => uniqueDate(
      event.classes
        .filter(c => c.class === eventClass)
        .map(c => c.date || event.startDate || new Date()),
    ),
  }
}

export const adminEventsAtom = atom<DecoratedEvent[]>({
  key: 'adminEvents',
  default: getEvents().then(events => events.map(decorateEvent)),
  effects: [
    logEffect,
    storageEffect,
  ],
})

export const adminEventByIdAtom = selectorFamily<DecoratedEvent | undefined, string>({
  key: 'adminEvents/eventId',
  get: (eventId) => async ({ get }) => {
    let event = get(adminEventsAtom).find(event => event.id === eventId)
    if (!event) {
      const fetched = await getEvent(eventId)
      if (fetched) {
        event = decorateEvent(fetched)
      }
    }
    return event
  },
  set: (eventId) => ({ set }, newValue) => {
    if (!newValue || newValue instanceof DefaultValue) {
      return
    }
    set(adminEventsAtom, oldEvents => {
      const index = oldEvents.findIndex(item => item.id === eventId)
      const newEvents = oldEvents.map(event => ({...event}))
      newEvents.splice(index, 1, newValue)
      return newEvents
    })
  },
})

export const adminEventSelector = selectorFamily<DecoratedEvent | undefined, string>({
  key: 'adminEvent',
  get: (eventId) => ({get}) => get(adminEventByIdAtom(eventId)),
  set: (eventId) => ({set}, newValue) => set(adminEventByIdAtom(eventId), newValue),
})

export const eventClassAtom = atom<string | undefined>({
  key: 'eventClass',
  default: selector({
    key: 'eventClass/default',
    get: ({ get }) => get(currentAdminEvent)?.uniqueClasses?.[0],
  }),
  effects: [
    logEffect,
    storageEffect,
  ],
})
