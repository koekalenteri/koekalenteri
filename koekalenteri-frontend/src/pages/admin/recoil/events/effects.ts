import { Event } from 'koekalenteri-shared/model'
import { AtomEffect } from 'recoil'

import { getEvents } from '../../../../api/event'

let loaded = false

export const remoteAdminEventsEffect: AtomEffect<Event[]> = ({ setSelf, trigger }) => {
  if (trigger === 'get' && !loaded) {
    loaded = true
    getEvents().then(setSelf)
  }
}
