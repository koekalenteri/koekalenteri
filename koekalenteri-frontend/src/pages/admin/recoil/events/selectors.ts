import type { Event } from 'koekalenteri-shared/model'

import { startOfToday } from 'date-fns'
import i18next from 'i18next'
import { DefaultValue, selector, selectorFamily } from 'recoil'

import { uniqueFn } from '../../../../utils'

import {
  adminEventFilterTextAtom,
  adminEventIdAtom,
  adminEventsAtom,
  adminShowPastEventsAtom,
  newEventAtom,
} from './atoms'

export const adminEventSelector = selectorFamily<Event | undefined, string | undefined>({
  key: 'adminEventSelector',
  get:
    (eventId) =>
    ({ get }) => {
      if (!eventId) {
        return get(newEventAtom)
      }
      const events = get(adminEventsAtom)
      return events.find((e) => e.id === eventId)
    },
  set:
    (eventId) =>
    ({ get, set }, value) => {
      if (!value || value instanceof DefaultValue) {
        return
      }
      const events = get(adminEventsAtom)
      const index = events.findIndex((e) => e.id === eventId)
      const newEvents = [...events]
      const insert = index === -1
      newEvents.splice(insert ? newEvents.length : index, insert ? 0 : 1, value)
      set(adminEventsAtom, newEvents)
    },
})

export const currentAdminEventSelector = selector({
  key: 'currentAdminEvent',
  get: ({ get }) => {
    const eventId = get(adminEventIdAtom)
    return eventId ? get(adminEventSelector(eventId)) : undefined
  },
  set: ({ get, set }, newValue) => {
    if (!newValue || newValue instanceof DefaultValue) {
      return
    }
    const eventId = get(adminEventIdAtom)
    if (!eventId) {
      return
    }
    set(adminEventSelector(eventId), newValue)
  },
})

export const filteredAdminEventsSelector = selector({
  key: 'filteredAdminEvents',
  get: ({ get }) => {
    const events = get(adminEventsAtom)
    const filter = get(adminEventFilterTextAtom).toLocaleLowerCase(i18next.language)
    const showPast = get(adminShowPastEventsAtom)

    return events.filter((event) => {
      return (
        !event.deletedAt &&
        (showPast || !event.startDate || event.startDate >= startOfToday()) &&
        (!filter ||
          [event.eventType, event.name, event.location, event.official?.name, event.secretary?.name]
            .join(' ')
            .toLocaleLowerCase(i18next.language)
            .includes(filter))
      )
    })
  },
})

export const adminEventOrganizersSelector = selector({
  key: 'adminEventOrganizers',
  get: ({ get }) => {
    const events = get(adminEventsAtom)
    return uniqueFn(
      events.map((event) => event.organizer),
      (a, b) => a.id === b.id
    ).sort((a, b) => a.name.localeCompare(b.name, i18next.language))
  },
})
