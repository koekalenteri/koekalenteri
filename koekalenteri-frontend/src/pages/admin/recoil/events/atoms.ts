import { addDays, nextSaturday, startOfDay, sub } from 'date-fns'
import { Event } from 'koekalenteri-shared/model'
import { atom, atomFamily, selector } from 'recoil'

import { uniqueClasses } from '../../../../utils'
import { logEffect, storageEffect } from '../../../recoil'

import { remoteAdminEventsEffect } from './effects'
import { adminEventSelector, currentAdminEventSelector } from './selectors'

export const adminEventsAtom = atom<Event[]>({
  key: 'adminEvents',
  default: [],
  effects: [
    logEffect,
    storageEffect,
    remoteAdminEventsEffect,
  ],
})

export const newEventStartDate = startOfDay(nextSaturday(addDays(Date.now(), 90)))
export const newEventEntryStartDate = sub(newEventStartDate, { weeks: 6 })
export const newEventEntryEndDate = sub(newEventStartDate, { weeks: 3 })

export const newEventAtom = atom<Event>({
  key: 'newEvent',
  default: {
    state: 'draft',
    startDate: newEventStartDate,
    endDate: newEventStartDate,
    entryStartDate: newEventEntryStartDate,
    entryEndDate: newEventEntryEndDate,
    classes: [],
    judges: [],
  } as unknown as Event,
  effects: [
    logEffect,
    storageEffect,
  ],
})

export const adminShowPastEventsAtom = atom<boolean>({
  key: 'adminShowPastEvents',
  default: false,
  effects: [
    logEffect,
    storageEffect,
  ],
})

export const adminEventFilterTextAtom = atom<string>({
  key: 'adminEventFilterText',
  default: '',
  effects: [
    logEffect,
    storageEffect,
  ],
})

export const adminEventIdAtom = atom<string | undefined>({
  key: 'adminEventId',
  default: undefined,
  effects: [
    logEffect,
    storageEffect,
  ],
})

export const eventClassAtom = atom<string | undefined>({
  key: 'eventClass',
  default: selector({
    key: 'eventClass/default',
    get: ({ get }) => {
      const event = get(currentAdminEventSelector)
      return uniqueClasses(event)[0] ?? event?.eventType
    },
  }),
  effects: [
    logEffect,
    storageEffect,
  ],
})

/**
 * Existing event editing, edits stored to local storage
 */
export const editableEventByIdAtom = atomFamily<Event | undefined, string>({
  key: 'editableEvent/Id',
  default: adminEventSelector,
  effects: [
    logEffect,
    storageEffect,
  ],
})
