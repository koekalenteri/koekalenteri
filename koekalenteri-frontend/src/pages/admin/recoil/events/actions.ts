import { useTranslation } from "react-i18next"
import { useNavigate } from "react-router-dom"
import { useAuthenticator } from "@aws-amplify/ui-react"
import { Event } from "koekalenteri-shared/model"
import { useSnackbar } from "notistack"
import { useRecoilState, useSetRecoilState } from "recoil"

import { putEvent } from "../../../../api/event"
import { Path } from "../../../../routeConfig"

import { adminEventIdAtom, newEventAtom } from "./atoms"
import { currentAdminEventSelector } from "./selectors"


export const useAdminEventActions = () => {
  const { user } = useAuthenticator(context => [context.user])
  const setAdminEventId = useSetRecoilState(adminEventIdAtom)
  const [currentAdminEvent, setCurrentAdminEvent] = useRecoilState(currentAdminEventSelector)
  const setNewEvent = useSetRecoilState(newEventAtom)
  const { enqueueSnackbar } = useSnackbar()
  const { t } = useTranslation()
  const navigate = useNavigate()

  return {
    copyCurrent,
    deleteCurrent,
    save,
  }

  function copyCurrent() {
    if (!currentAdminEvent) {
      return
    }
    const copy = structuredClone(currentAdminEvent)
    copy.id = ''
    copy.state = 'draft'
    copy.entries = 0
    copy.classes.forEach(c => c.entries = c.members = 0)
    delete copy.kcId

    setNewEvent(copy)
    navigate(Path.admin.newEvent)
  }

  async function save(event: Partial<Event>): Promise<Event> {
    const saved = await putEvent(event, user.getSignInUserSession()?.getIdToken().getJwtToken())
    setAdminEventId(saved.id)
    setCurrentAdminEvent(saved)
    return saved
  }

  async function deleteCurrent() {
    if (!currentAdminEvent || currentAdminEvent.deletedAt) {
      return
    }

    await save({
      ...currentAdminEvent,
      deletedAt: new Date(),
      deletedBy: user.attributes?.name || user.attributes?.email,
    })

    enqueueSnackbar(t('deleteEventComplete'), { variant: 'info' })
  }

}
