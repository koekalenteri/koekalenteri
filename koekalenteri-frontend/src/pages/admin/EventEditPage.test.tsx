import { Suspense } from 'react'
import { useTranslation } from 'react-i18next'
import { RouteObject } from 'react-router-dom'
import { ThemeProvider } from '@mui/material'
import { LocalizationProvider } from '@mui/x-date-pickers'
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFns'
import { render } from '@testing-library/react'
import { SnackbarProvider } from 'notistack'
import { RecoilRoot } from 'recoil'

import { eventWithStaticDates } from '../../__mockData__/events'
import theme from '../../assets/Theme'
import { Language, locales } from '../../i18n'
import { Path } from '../../routeConfig'
import { DataMemoryRouter, flushPromisesAndTimers } from '../../test-utils/utils'

import EventEditPage from './EventEditPage'

jest.useFakeTimers()

jest.mock('../../api/event')
jest.mock('../../api/eventType')
jest.mock('../../api/judge')
jest.mock('../../api/official')
jest.mock('../../api/organizer')
jest.mock('../../api/registration')

describe('EventEditPage', () => {
  it('renders properly', async () => {
    const { i18n } = useTranslation()
    const language = i18n.language as Language

    const routes: RouteObject[] = [{
      path: Path.admin.editEvent(),
      element: <EventEditPage />,
    }]

    const { container } = render(
      <ThemeProvider theme={theme}>
        <LocalizationProvider dateAdapter={AdapterDateFns} adapterLocale={locales[language]}>
          <RecoilRoot>
            <Suspense fallback={<div>loading...</div>}>
              <SnackbarProvider>
                <DataMemoryRouter initialEntries={[Path.admin.editEvent(eventWithStaticDates.id)]} routes={routes} />
              </SnackbarProvider>
            </Suspense>
          </RecoilRoot>
        </LocalizationProvider>
      </ThemeProvider>,
    )
    await flushPromisesAndTimers()
    expect(container).toMatchSnapshot()
  })
})
