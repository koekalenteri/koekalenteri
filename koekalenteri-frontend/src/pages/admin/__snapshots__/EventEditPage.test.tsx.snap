// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventEditPage renders properly 1`] = `
<div>
  <div>
    loading...
  </div>
</div>
`;
